Store Pending Deletions Fix

henningandersen opened the pull request:
FilterDirectory.getPendingDeletions does not delegate, fixed
temporarily by overriding in StoreDirectory.

This in turn caused duplicate file name use after a trimUnsafeCommits
had been done, since a new IndexWriter would not consider the pending
deletes in IndexFileDeleter. This should only happen on windows (AFAIK).

Reenabled doing index updates for all tests using
IndexShardTests.indexOnReplicaWithGaps (which could fail due to above
when using mocked WindowsFS).

Follow-up to #40249

A follow-up should fix this in FilterDirectory.

henningandersen committed:
Store Pending Files Fix

FilterDirectory.getPendingDeletions does not delegate, fixed
temporarily by overriding in StoreDirectory.

This in turn caused duplicate file name use after a trimUnsafeCommits
had been done, since a new IndexWriter would not consider the pending
deletes in IndexFileDeleter. This should only happen on windows (AFAIK).

Reenabled doing index updates for all tests using
IndexShardTests.indexOnReplicaWithGaps (which could fail due to above
when using mocked WindowsFS).

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch reviewed (approved):
Great catch! LGTM

dnhatn commented:
@henningandersen I also dug into a recent test failure of testRestoreLocalHistoryFromTranslogOnPromotion (https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos-6/312/console). I think that test failure has the same cause as the one you are fixing here. I proposed https://issues.apache.org/jira/browse/LUCENE-8734 to record nextWriteDVGen.

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample

s1monw commented on server/src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -760,6 +760,13 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
+        public Set<String> getPendingDeletions() throws IOException {
+            // FilterDirectory.getPendingDeletions does not delegate, working around it here.
+            // to be removed once fixed in FilterDirectory.
+            return unwrap(this).getPendingDeletions();
```
I think this fix is trappy, but I am ok with it for now. I would like us to delegate all our subclasses of `FilterDirectory` as well. We have a few more. I am torn on the unwrap to be honest.

henningandersen replied:
The problem with that route is that most of the subclasses of this are in lucene, including the `MockDirectoryWrapper` created by `LuceneTestCase` and for instance `HardlinkCopyDirectoryWrapper` used by `StoreRecovery`.
I agree that using `unwrap` is questionable, but in the given context this should be safe here until fixed in lucene.

s1monw replied:
I agree, we should still pro-actively implement the right fix in all subclasses and remove the unwrap once lucene is upgraded

henningandersen replied:
Thanks @s1monw  for clarifying. Except for two trivial test-only subclasses, I have added the override to all subclasses in 85f2066ec34b1e400d098d3ae0c742ee73288dfe.

Please have another look at your convenience.

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample

dnhatn reviewed (approved):
Great find!

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_lucene_directory_pending_files

henningandersen committed:
Store Pending Files Fix

Added getPendingDeletions delegation to all elasticsearch
FilterDirectory subclasses that were not trivial test-only overrides to
minimize the risk of hitting this issue in another case.
